pinot-controller/src/main/java/org/apache/pinot/controller/util/FileIngestionHelper.java

```diff
@@ -0,0 +1,140 @@
+ * Responsible for copying the file locally, building a segment and uploading it to the controller.
+ */
+public class FileIngestionHelper {
+  private static final Logger LOGGER = LoggerFactory.getLogger(FileIngestionHelper.class);
+
+  private static final String WORKING_DIR_PREFIX = "working_dir";
+  private static final String INPUT_DATA_DIR = "input_data_dir";
+  private static final String OUTPUT_SEGMENT_DIR = "output_segment_dir";
+  private static final String SEGMENT_TAR_DIR = "segment_tar_dir";
+  private static final String DATA_FILE_PREFIX = "data";
+
+  private final TableConfig _tableConfig;
+  private final Schema _schema;
+  private final BatchConfig _batchConfig;
+  private final ControllerConf _controllerConf;
+
+  public FileIngestionHelper(TableConfig tableConfig, Schema schema,
+      BatchConfig batchConfig, ControllerConf controllerConf) {
+    _tableConfig = tableConfig;
+    _schema = schema;
+    _batchConfig = batchConfig;
+    _controllerConf = controllerConf;
+  }
+
+  /**
+   * Creates a segment using the provided data file/URI and uploads to Pinot
+   */
+  public SuccessResponse buildSegmentAndPush(DataPayload payload)
+      throws Exception {
+    String tableNameWithType = _tableConfig.getTableName();
+
+    // Setup working dir
+    File workingDir = new File(FileUtils.getTempDirectory(),
+        String.format("%s_%s_%d", WORKING_DIR_PREFIX, tableNameWithType, System.currentTimeMillis()));
+    File inputDir = new File(workingDir, INPUT_DATA_DIR);
+    File outputDir = new File(workingDir, OUTPUT_SEGMENT_DIR);
+    File segmentTarDir = new File(workingDir, SEGMENT_TAR_DIR);
+    try {
+      Preconditions
+          .checkState(inputDir.mkdirs(), "Could not create directory for downloading input file locally: %s", inputDir);
+      Preconditions.checkState(segmentTarDir.mkdirs(), "Could not create directory for segment tar file: %s", inputDir);
+
+      // Copy file to local working dir
+      File inputFile =
+          new File(inputDir, String.format("%s.%s", DATA_FILE_PREFIX, _batchConfig.getInputFormat().toString().toLowerCase()));
+      if (payload._dataSource.equals(DataSource.URI)) {
+        FileIngestionUtils.copyURIToLocal(_batchConfig, payload._uri, inputFile);
+      } else {
+        FileIngestionUtils.copyMultipartToLocal(payload._multiPart, inputFile);
+      }
+
+      // Build segment
+      SegmentGeneratorConfig segmentGeneratorConfig =
+          FileIngestionUtils.generateSegmentGeneratorConfig(_tableConfig, _batchConfig, _schema, inputFile, outputDir);
+      String segmentName = FileIngestionUtils.buildSegment(segmentGeneratorConfig);
+
+      // Tar and push segment
+      File segmentTarFile =
+          new File(segmentTarDir, segmentName + org.apache.pinot.spi.ingestion.batch.spec.Constants.TAR_GZ_FILE_EXT);
+      TarGzCompressionUtils.createTarGzFile(new File(outputDir, segmentName), segmentTarFile);
+      FileIngestionUtils
```

Comment:
Anything that gets uploaded, will receive a unique segment name (tableName_mintime_maxtime_timestamp).
If a segment for a time range exists and user wishes to replace, they need to delete segment and upload again. Since this is meant as a quick test, it will be more common that users are uploading several files one by one, and they all have same time ranges. In this API we don't want the users to worry about providing a segment name generator etc, or doing some time column hacks to get unique names.